### PR TITLE
Improve type safety for client arguments in setup_call functions

### PR DESCRIPTION
**Pull request by koxudaxi**
Related Issues: https://github.com/Mirascope/mirascope/issues/513, #519

### CODE REVIEW

File: mirascope/core/base/_call_factory.py [hunk 1]
```diff
@@ -310,7 +490,7 @@ def base_call(
                 json_mode=json_mode,
                 client=client,
                 call_params=call_params,
-            )  # pyright: ignore [reportReturnType]
+            )  # pyright: ignore [reportReturnType, reportCallIssue]
```
Comment (koxudaxi) [inline mirascope/core/base/_call_factory.py:493]: I think I resolved all types. But, I can't resolve the error.
```bahs
  /Users/koudai/PycharmProjects/mirascope/mirascope/core/base/_call_factory.py:480:20 - error: No overloads for "decorator" match the provided arguments (reportCallIssue)
1 error, 0 warnings, 0 informations 
```

File: mirascope/core/base/_utils/_protocols.py [hunk 2]
```diff
@@ -33,6 +35,46 @@
...
+
+    @overload
+    def __call__(self, fn: Callable[_P, Messages.Type]) -> Callable[_P, _ResponseT]: ...
+
+    def __call__(
+        self, fn: Callable[_P, _BaseDynamicConfigT] | Callable[_P, Messages.Type]
+    ) -> Callable[_P, _ResponseT]: ...  # pragma: no cover
+
+
 class LLMFunctionDecorator(Protocol[_BaseDynamicConfigT, _ResponseT, _AsyncResponseT]):
```
Comment (willbakst) [inline mirascope/core/base/_utils/_protocols.py:78]: can we get rid of this if we have the sync vs. async and just use those now?

File: mirascope/core/base/_utils/_protocols.py [hunk 3]
```diff
@@ -108,12 +150,87 @@ def __call__(
...
+        CreateFn[_ResponseT, _ResponseChunkT]
+        | AsyncCreateFn[_ResponseT, _ResponseChunkT],
+        str | None,
+        list[Any],
+        list[type[_BaseToolT]] | None,
+        BaseCallKwargs,
+    ]: ...  # pragma: no cover
+
+
+class BaseClientSetupCall(
```
Comment (willbakst) [inline mirascope/core/base/_utils/_protocols.py:233]: why do we need both the base client setup call and the new protocol? can you explain what this solves so i know why we include both?

File: mirascope/core/base/_call_factory.py [hunk 4]
```diff
@@ -42,7 +47,9 @@
 _BaseCallParamsT = TypeVar("_BaseCallParamsT", bound=BaseCallParams)
 _BaseDynamicConfigT = TypeVar("_BaseDynamicConfigT", bound=BaseDynamicConfig)
 _BaseStreamT = TypeVar("_BaseStreamT", bound=BaseStream)
+_SyncBaseClientT = TypeVar("_SyncBaseClientT", bound=object)
```
Comment (willbakst) [inline mirascope/core/base/_call_factory.py:50]: nit: move this below the `_BaseClientT` line right below so this is right above it's async counterpart

File: mirascope/core/base/_utils/_protocols.py (hunk 3 above)
Comment (willbakst) [inline mirascope/core/base/_utils/_protocols.py:233]: we should name this so it's reason for existing is more clear

`SameSyncAndAsyncClientSetupCall` to it's more clear this is for providers who have a client that is shared across sync + async?

File: mirascope/core/base/_extract.py [hunk 5]
```diff
@@ -21,7 +22,9 @@
 from .tool import BaseTool
 
 _BaseCallResponseT = TypeVar("_BaseCallResponseT", bound=BaseCallResponse)
+_SyncBaseClientT = TypeVar("_SyncBaseClientT", bound=object)
 _BaseClientT = TypeVar("_BaseClientT", bound=object)
```
Comment (willbakst) [inline mirascope/core/base/_extract.py:26]: we could probably then also rename this to `_SameSyncAndAsyncClientT`

File: mirascope/core/base/_call_factory.py (hunk 1 above)
Comment (willbakst) [reply mirascope/core/base/_call_factory.py:515]: ok, that's fine for now, we can revisit this later if necessary

File: mirascope/core/base/_utils/_protocols.py
Comment (willbakst) [inline mirascope/core/base/_utils/_protocols.py:1]: you'll need to update bedrock with this stuff as well, so merge that PR first, then sync here and i'll approve

Comment (willbakst) [reply mirascope/core/base/_utils/_protocols.py:1]: ok i merged that PR in, so this should be ready for the updates

**Review by willbakst — APPROVED** [review]

**Pull request merged**